Moved bridge configuration setup within the operator

ppatierno opened the pull request:
### Type of change

- Refactoring

### Description

This PR fixes #10959 by moving the configuration of the bridge to be generated by the operator within a ConfigMap (actually using the same one that today is used for the metrics and logging configuratio, but with a different entry in the data section) then mounted on the bridge pod and loaded from there.

NOTE: the temporary change in the .azure/templates/steps/system_test_general.yaml file is for using a custom bridge image which doesn't use its script to generate the configuration anymore. Of course the commit to the file will be reverted before merging the PR. It's there for running STs.

### Checklist

- [x] Write tests
- [x] Make sure all tests pass
- [x] Try your changes from Pod inside your Kubernetes and OpenShift cluster, not just locally
- [x] Reference relevant issue(s) and close them after merging
- [x] Update CHANGELOG.md

ppatierno commented:
This PR comes together with https://github.com/strimzi/strimzi-kafka-bridge/pull/963

ppatierno commented:
/azp list

see-quick commented:
/azp list

see-quick commented:
/azp run bridge

ppatierno commented:
/azp run bridge

ppatierno commented:
/azp run regression

scholzj reviewed (commented):
i made just a quick pass ... but left bunch of comments.

ppatierno commented on systemtest/src/test/java/io/strimzi/systemtest/bridge/HttpBridgeST.java:
```diff
@@ -170,118 +167,6 @@ void testReceiveSimpleMessage() {
-            @Label(TestDocsLabels.BRIDGE)
-        }
-    )
-    void testCustomAndUpdatedValues() {
```
The test was about updating some env vars which are not used anymore because of the PR.

scholzj replied:
I don't think anything on what the test was doing changed. If anything, I'm pretty sure with this change this test needs to be extended to test the rollout of configuration changes separately.

scholzj replied:
This still applies.

ppatierno replied:
I have changed the way to check the configuration changes based on this PR now.

ppatierno commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaBridgeAssemblyOperator.java:
```diff
@@ -98,7 +98,7 @@ protected Future<KafkaBridgeStatus> createOrUpdate(Reconciliation reconciliation
             .compose(scale -> serviceOperations.reconcile(reconciliation, namespace, KafkaBridgeResources.serviceName(bridge.getCluster()), bridge.generateService()))
             .compose(i -> MetricsAndLoggingUtils.metricsAndLogging(reconciliation, configMapOperations, bridge.logging(), null))
-            .compose(metricsAndLogging -> configMapOperations.reconcile(reconciliation, namespace, KafkaBridgeResources.metricsAndLogConfigMapName(reconciliation.name()), bridge.generateMetricsAndLogConfigMap(metricsAndLogging)))
+            .compose(metricsAndLogging -> configMapOperations.reconcile(reconciliation, namespace, KafkaBridgeResources.metricsAndLogConfigMapName(reconciliation.name()), bridge.generateBridgeConfigMap(metricsAndLogging)))
```
Wdym?

scholzj replied:
What happens when the configuration changes? How do you reload it?

ppatierno replied:
I see what you mean and you are right. Before we were leveraging the changes on env vars applied to the Deployment in order to have the pod being rolled for us for free to restart the bridge and getting new config. Now there is a ConfigMap changing which will be updated at pod volume level at some point but bridge won't restart of course :-/ I will figure this out.

scholzj replied:
There isn't really anything to figure out ... do what we use elsewhere. Do a hash of the configuration when updating the config map and pass it as a Pod annotation to the Deployment. You just need to implement it. 😉

ppatierno replied:
I have "figured it out" and pushed ;-)

ppatierno commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaBridgeConfigurationBuilder.java:
```diff
@@ -0,0 +1,328 @@
+public class KafkaBridgeConfigurationBuilder {
+
+    // Names of environment variables expanded through config providers inside the bridge node
+    private static final String PLACEHOLDER_CERT_STORE_PASSWORD = "${strimzienv:CERTS_STORE_PASSWORD}";
+    private static final String PLACEHOLDER_SASL_USERNAME = "${strimzienv:KAFKA_BRIDGE_SASL_USERNAME}";
+    private static final String PASSWORD_VOLUME_MOUNT = "/opt/strimzi/bridge-password/";
+    // the SASL password file template includes: <volume_mount>/<secret_name>/<password_file>
+    private static final String PLACEHOLDER_SASL_PASSWORD_FILE_TEMPLATE = "${strimzidir:%s%s:%s}";
+    private static final String PLACEHOLDER_OAUTH_CONFIG = "${strimzienv:KAFKA_BRIDGE_OAUTH_CONFIG}";
+    private static final String PLACEHOLDER_OAUTH_ACCESS_TOKEN = "${strimzienv:KAFKA_BRIDGE_OAUTH_ACCESS_TOKEN}";
+    private static final String PLACEHOLDER_OAUTH_REFRESH_TOKEN = "${strimzienv:KAFKA_BRIDGE_OAUTH_REFRESH_TOKEN}";
+    private static final String PLACEHOLDER_OAUTH_CLIENT_SECRET = "${strimzienv:KAFKA_BRIDGE_OAUTH_CLIENT_SECRET}";
+    private static final String PLACEHOLDER_OAUTH_PASSWORD_GRANT_PASSWORD = "${strimzienv:KAFKA_BRIDGE_OAUTH_PASSWORD_GRANT_PASSWORD}";
```
But it would mean having string hardcoded where we use it. I would stick with constant here.

scholzj replied:
Doesn't that just make the code harder to read? You wanna see what the configuration will look like, but you actually have to jump between the definitions instead of having it inlined. As I said, I think it makes sense to have this for those that are used multiple times. For those used just once it seems like a waste of space.

ppatierno replied:
It's not a big issue jumping on a constant definition as weel as today, most of the IDEs shows the value of the constant straight away.

scholzj replied:
Right, works in your IDE, so whoever else can go f*ck themself. :-D

scholzj replied:
Maybe we could at least do it properly ... 
```java
jaasConfig.append(SPACE + OAUTH_PASSWORD_GRANT_PASSWORD_KEY + EQUALS + PLACEHOLDER_OAUTH_PASSWORD_GRANT_PASSWORD);
```
:-D

scholzj replied:
If you decide to stick with it, you should probably at least better name it to make it more obvious what it is. For example `PLACEHOLDER_SASL_PASSWORD_FILE_TEMPLATE` is completely unclear if that is file path, env var config provider, file config provider etc.

ppatierno replied:
Tbh I can't see your comments being funny if that was the intention.
Anyway the constants here are neither more nor less than the same we have in the `KafkaBrokerConfigurationBuilder` here https://github.com/strimzi/strimzi-kafka-operator/blob/main/cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaBrokerConfigurationBuilder.java#L61

scholzj replied:
Nope, that was not the intention. There were meant seriously.

scholzj replied:
> Anyway the constants here are neither more nor less than the same we have in the KafkaBrokerConfigurationBuilder here https://github.com/strimzi/strimzi-kafka-operator/blob/main/cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaBrokerConfigurationBuilder.java#L61

Is there a placeholder that is used **once**?

ppatierno replied:
Nope (even if 2 is parent of 1) but names are anyway the same not making any sense as well.

scholzj replied:
So, are you saying that because two variable names in other class are bad/non-descriptive we should make all variable names bad/non-descriptive? Or does it apply only because I named them badly?

ppatierno replied:
No I am saying that for me the names are good on both classes. But from your perspective I would say that if they are bad in the bridge configuration class, also the broker configuration class has bad constant names (which are from where I got inspiration because, again, I find them understandable).
Anyway, I am open to specific suggestions to fix them both.

scholzj replied:
TBH, I think it is easier to live with the names as long as they are used a few times and not for tens of things used just once. I think they can be easily improved by just making them more descriptive. E.g. `PLACEHOLDER_CERT_STORE_PASSWORD` could be `PLACEHOLDER_CERT_STORE_PASSWORD_CONFIG_PROVIDER_FROM_ENV_VAR` or something like that.

ppatierno replied:
I renamed them in the bridge configuration builder. If the names are ok, I will open a separate PR to address the same in the broker configuration builder.

scholzj replied:
I still don't get this. You haven't given a single reason why having a special variable that is used once adds some value.

ppatierno replied:
Too be honest I don't see it really appealing having some hard coded constants and other declared at the top of the file (just because used in one more place). To be honest I would stick with consistency for them all. Is it really a so big problem? Even reviewing on GitHub with the "search" function it's not that difficult to get the constant value imho. You do that anyway when the constant is used in multiple place and you need to know the value. So from this pov it doesn't make any difference.

scholzj replied:
> Too be honest I don't see it really appealing having some hard coded constants and other declared at the top of the file (just because used in one more place). To be honest I would stick with consistency for them all.

What is the actual benefit of that? What does it make easier? Also, what consistency are you trying to achieve? You have plenty of hardcoded values there: 
*`PKCS12` is hardcoded
* The security protocol strings are hardcoded
* SASL mechanisms are hardcoded
* Rack ID is hardcoded
* Provider class names are hardcoded
* The configuration keys for all the options are all hardcoded

So what exactly is the underlying logic that makes this consistent?

Please help me to understand it. Because I feel like I gave a pretty reasonable comment that helps improve the code readability for more than just myself and helps to eliminate any need for searching for things or using IDE to read the code. And I do not really see any downside we will get when we apply it. But instead of explaining the benefits of your solution, you seem to keep telling me to use IDE or search. So if consistency is supposed to be the benefit, what exactly is the pattern you are trying to be consistent with?

> Even reviewing on GitHub with the "search" function it's not that difficult to get the constant value imho. You do that anyway when the constant is used in multiple place and you need to know the value. So from this pov it doesn't make any difference.

Yeah, try that for example with the next PR when the constant definition is hidden because nothing around it changed but the use of it has changed. Also, even without that, it still means that you need to search for it instead of just seeing it. So what is the positive value that stands against it to justify it?

ppatierno replied:
Let's move forward. I removed the constants used just in one place and hard coded the strings. I will push the changes soon.

ppatierno commented on cluster-operator/src/test/java/io/strimzi/operator/cluster/model/KafkaBridgeConfigurationBuilderTest.java:
AFAIU, `IsEquivalent` matches all the lines. I don't want to check the entire configuration each time for each test but focusing on the parameters I expect for that specific test (i.e. the ssl trustore, ssl keystore, sasl, ...). With `IsEquivalent` I should always set, for example, bridge id, bootstrap servers as expected lines.

scholzj replied:
If you look at how it is used in the Broker config, you will see that there it elverages the builder pattern to basically check only the particular segment with minimal overhead (which is basically just `node.id=...`). Why is this not usable here?

ppatierno replied:
I see what you mean but I also found something weird when looking at it in the `KafkaBrokerConfigurationBuilder`. But let's start from my doubt ...
As you can see, I am setting up the config providers in the `KafkaBridgeConfigurationBuilder` ctor so it means that using `IsEquivalent` I have to put all the 7 lines generated by the `configureConfigProviders()` method in the expected lines otherwise the test fails (even if I want to test just bridge id and bootstrap servers). 
This led me to understand how it was working with broker configuration and I noticed that the config providers are set only when `withUserConfiguration` method is called but not when, for example, `withRackId` is called which needs config providers in the end. The same `withCruiseControl`, it would need config providers. So in the broker config test, when testing the rack id you are not testing that config providers is there (which should be because needed for the configuration to work) ... is it right? At runtime everything works fine because the `withUserConfiguration` is always called and it sets the config providers (for the rack, for CC, etc etc ...).
This is why I set the config providers in the constructor because they are needed across several `withXXXX` methods (but of course they could end in the configuration even if not needed).

scholzj replied:
> This led me to understand how it was working with broker configuration and I noticed that the config providers are set only when withUserConfiguration method is called but not when, for example, withRackId is called which needs config providers in the end. The same withCruiseControl, it would need config providers. So in the broker config test, when testing the rack id you are not testing that config providers is there (which should be because needed for the configuration to work) ... is it right? At runtime everything works fine because the withUserConfiguration is always called and it sets the config providers (for the rack, for CC, etc etc ...).
This is why I set the config providers in the constructor because they are needed across several withXXXX methods (but of course they could end in the configuration even if not needed).

It is unit testing -> you split things into smaller units and test them individually. That is the whole idea. The confgi providers are setup in the user configuration because:
* Users might configure their own config providers (not sure if this case is there for the Bridge and if it needs to be handled)
* We know the user configuration is always called anyway -> you can test that with a single test in `KafkaBridgeClusterTest` with single configuration while testing the various options is covered on a smaller units in the `KafkaBridgeConfigurationBuilderTest`

> As you can see, I am setting up the config providers in the KafkaBridgeConfigurationBuilder ctor so it means that using IsEquivalent I have to put all the 7 lines generated by the configureConfigProviders() method in the expected lines otherwise the test fails (even if I want to test just bridge id and bootstrap servers).

That is indeed annoying. But you can split it into a separate `with*` methods instead of doing it in the constructor I guess?

----

There is one big advantage with `IsEqvivalent` -> It tests the exact content. The `containsString` checks that something is present, but not what else might be present. So you might not catch duplicate or invalid config options for example with the way you have the test now. (The has real meaning, it is eassentially the same as why for example the #11022 bug slipped in)

ppatierno replied:
> That is indeed annoying. But you can split it into a separate with* methods instead of doing it in the constructor I guess?

In fact this is the only solution I had in mind with a dedicated `withConfigProviders` method.
But I would argue that the same could be done for the broker because as I mentioned, even if just unit testing, when set rackid or CC, you would expect the needed config providers being there which is actually not tested instead because they are not set (unless you set the user configuration).

scholzj replied:
No, you do not care if some other method sets the config providers properly when you test `withRackId`. And yes, the broker configuration can probably have some method `withConfigProviders`. But I doubt it would help the code because as I said, you need to reconcile it with any user-set config providers. You might need to do that in the Bridge as well.

ppatierno commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaBridgeConfigurationBuilder.java:
Hopefully it's not better.

tinaselenge commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaBridgeConfigurationBuilder.java:
```diff
@@ -0,0 +1,340 @@
+     * @param tls   client TLS configuration
+     * @return  the builder instance
+     */
+    public KafkaBridgeConfigurationBuilder withTls(ClientTls tls/*, KafkaClientAuthentication authentication*/) {
```
Is the commented out `KafkaClientAuthentication authentication` intentional?

ppatierno replied:
Yeah it's a left over, thanks!

tinaselenge reviewed (approved):
LGTM, thanks :)

fvaleri reviewed (commented):
There are a couple of test failures, but the overall approach LGTM.

I also left a minor comment. Thanks.

scholzj commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaBridgeConfigurationBuilder.java:
```diff
@@ -0,0 +1,342 @@
+     *
+     * @return  the builder instance
+     */
+    public KafkaBridgeConfigurationBuilder withConfigProviders() {
+        printSectionHeader("Config providers");
+        writer.println("kafka.config.providers=strimzienv,strimzifile,strimzidir");
+        writer.println("kafka.config.providers.strimzienv.class=org.apache.kafka.common.config.provider.EnvVarConfigProvider");
+        writer.println("kafka.config.providers.strimzienv.param.allowlist.pattern=.*");
+        writer.println("kafka.config.providers.strimzifile.class=org.apache.kafka.common.config.provider.FileConfigProvider");
+        writer.println("kafka.config.providers.strimzifile.param.allowed.paths=/opt/strimzi");
+        writer.println("kafka.config.providers.strimzidir.class=org.apache.kafka.common.config.provider.DirectoryConfigProvider");
+        writer.println("kafka.config.providers.strimzidir.param.allowed.paths=/opt/strimzi");
+        writer.println();
+        return this;
+    }
```
If we don't allow users to configure the providers, it should be mentioned in the release notes as that is potentially backwards incompatible. But maybe we can simply handle it and just add our own as we do for broker?

ppatierno replied:
Yeah I changed to add the support as we have for broker.

scholzj commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaBridgeAssemblyOperator.java:
```diff
@@ -98,10 +104,17 @@ protected Future<KafkaBridgeStatus> createOrUpdate(Reconciliation reconciliation
-            .compose(metricsAndLogging -> configMapOperations.reconcile(reconciliation, namespace, KafkaBridgeResources.metricsAndLogConfigMapName(reconciliation.name()), bridge.generateMetricsAndLogConfigMap(metricsAndLogging)))
+            .compose(metricsAndLogging -> {
+                ConfigMap configMap = bridge.generateBridgeConfigMap(metricsAndLogging);
+                podAnnotations.put(KafkaBridgeCluster.ANNO_STRIMZI_IO_CONFIGURATION_HASH, Util.hashStub(configMap.getData().get(KafkaBridgeCluster.BRIDGE_CONFIGURATION_FILENAME)));
```
This looks good. But it might make sense to have some tests to check it is set / changes?

scholzj commented on cluster-operator/src/test/java/io/strimzi/operator/cluster/model/KafkaBridgeClusterTest.java:
```diff
@@ -283,6 +273,11 @@ public void testGenerateDeploymentWithTls() {
         assertThat(io.strimzi.operator.cluster.TestUtils.containerEnvVars(containers.get(0)).get(KafkaBridgeCluster.ENV_VAR_KAFKA_BRIDGE_TLS), is("true"));
+
+        ConfigMap configMap = kbc.generateBridgeConfigMap(metricsAndLogging);
+        assertThat(configMap.getData().get(KafkaBridgeCluster.BRIDGE_CONFIGURATION_FILENAME), containsString("kafka.ssl.truststore."));
+        assertThat(configMap.getData().get(KafkaBridgeCluster.BRIDGE_CONFIGURATION_FILENAME), not(containsString("kafka.ssl.keystore.")));
+        assertThat(configMap.getData().get(KafkaBridgeCluster.BRIDGE_CONFIGURATION_FILENAME), containsString("kafka.security.protocol=SSL"));
```
Why should it contain / not contain these values?

ppatierno replied:
Because the test is enabling TLS (with trusted certificates) on the bridge so I am verifying that the corresponding generated configuration contains what to expect (not to expect, i.e. keystore because no TLS authentication configured).

scholzj replied:
Would it make more sense to have the whole file here once and otherwise rely on the unit tests of the configuration builder?

ppatierno replied:
Yeah I will just remove this checks because we already test these things in the unit test. Also, before the changes this part was just checking one TLS env vars so it was anyway not checking the full configuration.

scholzj commented on cluster-operator/src/test/java/io/strimzi/operator/cluster/model/KafkaBridgeClusterTest.java:
```diff
@@ -314,6 +309,11 @@ public void testGenerateDeploymentWithTlsAuth() {
         assertThat(io.strimzi.operator.cluster.TestUtils.containerEnvVars(containers.get(0)).get(KafkaBridgeCluster.ENV_VAR_KAFKA_BRIDGE_TLS), is("true"));
+
+        ConfigMap configMap = kbc.generateBridgeConfigMap(metricsAndLogging);
+        assertThat(configMap.getData().get(KafkaBridgeCluster.BRIDGE_CONFIGURATION_FILENAME), containsString("kafka.ssl.truststore."));
+        assertThat(configMap.getData().get(KafkaBridgeCluster.BRIDGE_CONFIGURATION_FILENAME), containsString("kafka.ssl.keystore."));
+        assertThat(configMap.getData().get(KafkaBridgeCluster.BRIDGE_CONFIGURATION_FILENAME), containsString("kafka.security.protocol=SSL"));
```
Same as above, why should it / should not contain these values?

ppatierno replied:
The same as before, with in addition the keystore configuration because the TLS authentication is enabled as well.

scholzj commented on cluster-operator/src/test/java/io/strimzi/operator/cluster/model/KafkaBridgeClusterTest.java:
```diff
@@ -970,44 +970,6 @@ public void testNullClusterRoleBinding() {
     }
 
-    @ParallelTest
-    public void testKafkaBridgeContainerEnvVarsConflict() {
```
We still have some env vars, so why not update it instead of deleting it?

ppatierno replied:
I brought it back changing the env vars to be used in the check.

tinaselenge commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaBridgeCluster.java:
```diff
@@ -411,45 +407,6 @@ protected List<EnvVar> getEnvVars() {
-        varList.add(ContainerUtils.createEnvVar(ENV_VAR_KAFKA_BRIDGE_ID, cluster));
-
-        if (kafkaBridgeConsumer != null) {
-            varList.add(ContainerUtils.createEnvVar(ENV_VAR_KAFKA_BRIDGE_CONSUMER_CONFIG, new KafkaBridgeConsumerConfiguration(reconciliation, kafkaBridgeConsumer.getConfig().entrySet()).getConfiguration()));
```
Can KafkaBridgeConsumerConfiguration be deleted since it was only used here? The same goes for KafkaBridgeProducerConfiguration.

scholzj replied:
That seems wrong. That is how user passes the configuration. Those classes should be used.

tinaselenge replied:
Oh I see, I was wondering where these model classes are used but these are used by the users. Please ignore this comment.

scholzj replied:
No, what I meant is that I think they need to be used in this PR -> the process the user configuration from the KafkaBridge CR and should be used to pass it into the config file here. What I was trying to say is that this likely means some bug that should be fixed rather than that the files should be deelted.

tinaselenge replied:
Ok, so that's what I was wondering in the first place. Perhaps these should be used in the config builder class similar to https://github.com/strimzi/strimzi-kafka-operator/blob/c1b20f726dddbcd2a070c2eeb14fd30902027aec/cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaBrokerConfigurationBuilder.java#L478?

scholzj replied:
Yeah. Or maybe this is more corresponding: https://github.com/strimzi/strimzi-kafka-operator/blob/c1b20f726dddbcd2a070c2eeb14fd30902027aec/cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaBrokerConfigurationBuilder.java#L795

scholzj replied:
That should not work that way. The configuration classes are what ensures the filtering of forbidden options, defaults etc. They are used everywhere else and I think we should stick with that.

ppatierno replied:
You are right. I will fix this.

tinaselenge commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaBridgeCluster.java:
```diff
@@ -411,45 +407,6 @@ protected List<EnvVar> getEnvVars() {
-
         if (tls != null) {
             varList.add(ContainerUtils.createEnvVar(ENV_VAR_KAFKA_BRIDGE_TLS, "true"));
 
```
Do we still need this env var since we now have `withTls()` method that checks if tls is null?

ppatierno replied:
actually you are correct, we don't need this one but still need the trusted certs one because of the TLS certs prepartion script in the bridge here https://github.com/strimzi/strimzi-kafka-bridge/blob/main/bin/docker/kafka_bridge_tls_prepare_certificates.sh#L56

tinaselenge commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaBridgeConfigurationBuilder.java:
```diff
@@ -0,0 +1,342 @@
+    /**
+     * Adds the TLS/SSL configuration for the bridge to connect to the Kafka cluster.
+     * The configuration includes the trusted certificates store for TLS connection (server authentication)
+     * as well as the keystore configuration for mTLS (client authentication).
```
this method seems to only configure truststore, but not the keystore as the doc suggests

ppatierno replied:
Good catch! It's a leftover when the method did that, but not anymore because the logic is all within the `withAuthentication` method.

ppatierno commented:
/azp run regression

scholzj commented on cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/assembly/KafkaBridgeAssemblyOperatorTest.java:
```diff
@@ -167,8 +164,15 @@ public void testCreateOrUpdateCreatesCluster(VertxTestContext context) {
-                assertThat(dc, is(bridge.generateDeployment(Collections.singletonMap(Annotations.ANNO_STRIMZI_AUTH_HASH, "0"), true, null, null)));
+                assertThat(dc, is(bridge.generateDeployment(Map.of(
+                        Annotations.ANNO_STRIMZI_AUTH_HASH, "0",
+                        KafkaBridgeCluster.ANNO_STRIMZI_IO_CONFIGURATION_HASH, Util.hashStub(cm.getData().get(KafkaBridgeCluster.BRIDGE_CONFIGURATION_FILENAME))
```
I would use a fixed value here. That helps to make sure the hash is _stable_.

scholzj commented on cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/assembly/KafkaBridgeAssemblyOperatorTest.java:
```diff
@@ -351,8 +340,15 @@ public void testCreateOrUpdateUpdatesCluster(VertxTestContext context) {
-                assertThat(dc, is(compareTo.generateDeployment(Collections.singletonMap(Annotations.ANNO_STRIMZI_AUTH_HASH, "0"), true, null, null)));
+                assertThat(dc, is(compareTo.generateDeployment(Map.of(
+                        Annotations.ANNO_STRIMZI_AUTH_HASH, "0",
+                        KafkaBridgeCluster.ANNO_STRIMZI_IO_CONFIGURATION_HASH, Util.hashStub(cm.getData().get(KafkaBridgeCluster.BRIDGE_CONFIGURATION_FILENAME))
```
Same as above. Does any of the test actually check that the value changes?

ppatierno replied:
I added a test for that.

scholzj commented on .azure/templates/steps/system_test_general.yaml:
```diff
@@ -145,7 +145,7 @@ jobs:
       env:
         DOCKER_TAG: $(docker_tag)
-        BRIDGE_IMAGE: "latest-released"
+        BRIDGE_IMAGE: "quay.io/ppatierno/kafka-bridge:bridge-config"
```
We need to address this before merging this.

ppatierno replied:
Yes sure.

ppatierno commented on CHANGELOG.md:
```diff
@@ -6,6 +6,7 @@
 * Support for MirrorMaker 1 has been removed
 * Added support to configure `dnsPolicy` and `dnsConfig` using the `template` sections.
 * Store Kafka node certificates in separate Secrets, one Secret per pod.
+* Moved HTTP bridge configuration to the ConfigMap setup by the operator.
```
I think it's not worth the effort of supporting both. I will add a line in the "Major changes, deprecations and removals" section

im-konge commented on systemtest/src/test/java/io/strimzi/systemtest/bridge/HttpBridgeST.java:
```diff
@@ -278,8 +268,14 @@ void testCustomAndUpdatedValues() {
+        Map<String, Object> producerConfigMap = config.entrySet().stream().filter(e -> e.getKey().startsWith("kafka.producer.")).collect(Collectors.toMap(e -> e.getKey().replace("kafka.producer.", ""), Map.Entry::getValue));
+        Map<String, Object> consumerConfigMap = config.entrySet().stream().filter(e -> e.getKey().startsWith("kafka.consumer.")).collect(Collectors.toMap(e -> e.getKey().replace("kafka.consumer.", ""), Map.Entry::getValue));
+        assertThat(producerConfigMap.entrySet().containsAll(producerConfig.entrySet()), is(true));
+        assertThat(consumerConfigMap.entrySet().containsAll(consumerConfig.entrySet()), is(true));
```
IIRC we don't have any special method for that, so this LGTM

ppatierno commented:
/azp run regression

ppatierno commented:
@scholzj I think I addressed all your comments.
Regarding the changes you see on the values.yaml and HelmResource.java, they are temporary for running the STs and allowing to bring the right operator and bridge for the Helm Charts one. I also opened an issue about that https://github.com/strimzi/strimzi-kafka-operator/issues/11066
Of course these changes will be reverted before merging the PR.

ppatierno commented:
actually it was worse. With the above changes I was able to run HelmCharts system tests locally but now the build is failing because of some checks we do on the naming:

```shell
FAIL  strimzi operator deployment	strimzi-kafka-operator/tests/operator_deployment_test.yaml
	- should have latest image with default image registry when no image is provided

		- asserts[1] `matchRegex` fail
			Template:	strimzi-kafka-operator/templates/060-Deployment-strimzi-cluster-operator.yaml
			DocumentIndex:	0
			Path:	spec.template.spec.containers[0].image
			Expected to match:
				quay.io/strimzi/operator:(latest|[0-9]+\.[0-9]+\.[0-9]+)
			Actual:
				quay.io/ppatierno/operator:bridge-config
```

Tbh, there are too many changes to make for having all regression STs running as they are now. What should we do? @scholzj @im-konge ideas?

ppatierno commented:
/azp run regression

ppatierno commented:
/azp run regression

ppatierno commented:
So as mentioned on the related bridge PR https://github.com/strimzi/strimzi-kafka-bridge/pull/963, I made a change on the `bridge.version` to set "latest" and run `make bridge_version` and then push the changes.
This way all the needed deployment files were updated for running the STs and they are all green. Of course, we have to revert this change as soon as we have a bridge 0.32.0 release. I will open an issue for that when this PR is merged.
@scholzj would you mind to have another pass and checking if everything looks fine to you now?

scholzj commented on CHANGELOG.md:
```diff
@@ -21,6 +22,8 @@
   If you are using the storage overrides, you should instead use multiple KafkaNodePool resources with a different storage class each.
   For more details about migrating from storage overrides, please follow the [documentation](https://strimzi.io/docs/operators/0.45.0/full/deploying.html#con-config-storage-zookeeper-str).
+* Because of the corresponding configuration changes, the 0.32.0 is the minimum HTTP bridge release working with the Strimzi operator 0.46.0.
+  * If you have an Apache Kafka cluster running together with an HTTP bridge instance, please upgrade the bridge after upgrading the operator.
```
I think this part is confusing -> the Bridge has no separate versioning when deployed through the operator and there is no way to upgrade the Bridge after upgrading the operator. For most users, all of this happens automatically. This might be relevant only to someone with heavily customized operator installation. I wonder if the previous line is sufficient for that and those using customized installation would understand it. But if you think that is not sufficient, we definitely need to clarify this part to not confuse anyone.

ppatierno replied:
You are right. You can't upgrade the bridge, it's coming with the operator installation. I will remove this sentence.

scholzj reviewed (approved):
One nit for the CHANGELOG. LGTM otherwise.
